roles/translator/src/template_receiver/mod.rs

```diff
@@ -0,0 +1,142 @@
+use codec_sv2::{StandardEitherFrame, StandardSv2Frame};
+use roles_logic_sv2::utils::Mutex;
+
+use codec_sv2::Frame;
+use roles_logic_sv2::{
+    handlers::{template_distribution::ParseServerTemplateDistributionMessages, SendTo_},
+    parsers::{PoolMessages, TemplateDistribution},
+    template_distribution_sv2::{
+        CoinbaseOutputDataSize, NewTemplate, SetNewPrevHash, SubmitSolution,
+    },
+};
+pub type SendTo = SendTo_<roles_logic_sv2::parsers::TemplateDistribution<'static>, ()>;
+//use messages_sv2::parsers::JobNegotiation;
+pub type Message = PoolMessages<'static>;
+pub type StdFrame = StandardSv2Frame<Message>;
+pub type EitherFrame = StandardEitherFrame<Message>;
+use async_channel::{Receiver, Sender};
+use network_helpers::PlainConnection;
+use std::{convert::TryInto, net::SocketAddr, sync::Arc};
+mod message_handler;
+mod setup_connection;
+use setup_connection::SetupConnectionHandler;
+
+pub struct TemplateRx {
+    receiver: Receiver<EitherFrame>,
+    sender: Sender<EitherFrame>,
+    send_new_tp_to_negotiator: Sender<(NewTemplate<'static>, u64)>,
+    send_new_ph_to_negotiator: Sender<(SetNewPrevHash<'static>, u64)>,
+}
+
+impl TemplateRx {
+    pub async fn connect(
+        address: SocketAddr,
+        send_new_tp_to_negotiator: Sender<(NewTemplate<'static>, u64)>,
+        send_new_ph_to_negotiator: Sender<(SetNewPrevHash<'static>, u64)>,
+        receive_coinbase_output_max_additional_size: Receiver<(CoinbaseOutputDataSize, u64)>,
+        solution_receiver: Receiver<SubmitSolution<'static>>,
+    ) {
+        let stream = async_std::net::TcpStream::connect(address).await.unwrap();
+
+        let (mut receiver, mut sender): (Receiver<EitherFrame>, Sender<EitherFrame>) =
+            PlainConnection::new(stream, 10).await;
+
+        SetupConnectionHandler::setup(&mut receiver, &mut sender, address)
+            .await
+            .unwrap();
+
```

Comment:
I'd recommend some logging here - if for some reason the `recv` below never comes through we'd be blocked and not know why.